Source/CarthageKit/Project.swift

```diff
@@ -587,21 +587,65 @@ public final class Project { // swiftlint:disable:this type_body_length
 		pinnedVersion: PinnedVersion,
 		toolchain: String?
 	) -> SignalProducer<URL, CarthageError> {
+		
+		typealias FrameworkUrlAndDestinationUrl = (frameworkUrlInUnarchivedFolder: URL, frameworkUrlInBuildForlder: URL)
+
 		return SignalProducer<URL, CarthageError>(value: zipFile)
 			.flatMap(.concat, unarchive(archive:))
 			.flatMap(.concat) { directoryURL -> SignalProducer<URL, CarthageError> in
+				// For all frameworks in the directory where the archive has been expanded
 				return frameworksInDirectory(directoryURL)
-					.flatMap(.merge) { url -> SignalProducer<URL, CarthageError> in
-						return checkFrameworkCompatibility(url, usingToolchain: toolchain)
+					.collect()
+					// Check if multiple framewroks resolve to the same unique destination URL in the Carhage/Build/ folder.
+					// This is needed because frameworks might overwrite each others.
+					.flatMap(.merge) { frameworksUrls -> SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError> in
+						return SignalProducer<URL, CarthageError>(frameworksUrls)
+							.flatMap(.merge) { url -> SignalProducer<URL, CarthageError> in
+								return platformForFramework(url)
+									.attemptMap { platform -> Result<URL, CarthageError> in
+										guard let destinationURLInWorkingDir = platform
+											.relativeURL?
+											.appendingPathComponent(url.lastPathComponent, isDirectory: true) else {
+												return .failure(.internalError(description: "failed to construct framework destination url from \(platform) and \(url.lastPathComponent)"))
+										}
+
+										return .success(self
+											.directoryURL
+											.appendingPathComponent(destinationURLInWorkingDir.path, isDirectory: true))
+								}
+							}
+							.uniqueValues()
+							.collect()
+							.flatMap(.merge) { uniqueDestinationUrls -> SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError> in
+								guard uniqueDestinationUrls.count == frameworksUrls.count else {
+									return SignalProducer<FrameworkUrlAndDestinationUrl, CarthageError>(error: CarthageError.invalidArchive(description: "found multiple frameworks with the same unarchiving destination"))
```

Comment:
Ideally this error would say _which_ frameworks have the same destination

Reply:
Yeah... I was lazy 😅  I'll try to make it smarter